### PR TITLE
Add rule S1656 (based on 'no-self-assign') for TS/JS

### LINKED ISSUE #1427: Add rule S1656 (based on 'no-self-assign') for TS/JS

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #1427 

Reason for differences
* eslint implementation doesn't consider mutating functions on arrays `sort` and `reverse`, ie. `a = a.sort()` is not raised
* eslint implementations detects equivalent expressions which differ due to parenthesis `a = (a)`
* `no-self-assign` doesn't handle member expressions using `this`, so  `this.x = this.x` is not detected

### CODE REVIEW

File: its/ruling/src/test/expected/ts/vscode/typescript-S1656.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+{
+'vscode:build/lib/nls.ts':[
+355,
+],
+'vscode:src/vs/base/browser/ui/list/listWidget.ts':[
+664,
```
Comment (vilchik-elena) [inline its/ruling/src/test/expected/ts/vscode/typescript-S1656.json]: you are saying array mutating methods are not reported but these issues are exactly that

Comment (saberduck) [reply its/ruling/src/test/expected/ts/vscode/typescript-S1656.json]: my bad, expectations were not updated correctly, that's why QA failed. These issues are not raised anymore

File: its/ruling/src/test/expected/ts/angular/typescript-S1656.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+'angular:packages/compiler-cli/src/path_mapped_compiler_host.ts':[
+134,
```
Comment (vilchik-elena) [inline its/ruling/src/test/expected/ts/angular/typescript-S1656.json]: `sf.fileName = sf.fileName;` here and you say that we don't report on properties

Comment (saberduck) [reply its/ruling/src/test/expected/ts/angular/typescript-S1656.json]: indeed, I was wrong, what's not working is detecting self assignment with `ThisExpression` as can be seen in https://github.com/eslint/eslint/blob/master/lib/rules/no-self-assign.js#L66 , maybe we can contribute a fix for that ( `lobj.type` is `ThisExpression` )

**Pull request merged**